### DISCUSSION TITLE
Maintaining LLVM passes for CMPLOG

### DISCUSSION THREAD

**Discussion by TheCoryBarker** [AFLplusplus/AFLplusplus#1569] [General]
Hi maintainers,

I am wondering how often LLVM updates break the passes needed for CMPLOG, and the code needs to be updated? Is there frequent maintenance here? And is it much work to make these updates?

I was pointed to this line in the Changelog:
"trying to stay on top on all these #$&§!! changes in llvm 15"

which makes me think there is a lot of work involved here.

Thank you!

**Comment by vanhauser-thc** [comment]
Some of the issues are our problem, llvm changes the api a lot, and although they mark stuff as deprecated and wait for 2 versions until stuff is removed we several times lost track of this and the removal then broke plugins.

my gripes are though that there are changes from time to time that break compilation but are unannounced - maybe because they are bugs?  
 and then the code quality gets worse and worse, e.g. loading plugins into lld was switched from old format to new without any transition time, and then also bugged so plugins could not be loaded at all, and it tool several months to fix that.

finally, support is really bad. yes the documentation online is ok, and you can read the code, but sometimes things dont work or dont compile, or you want to do something and it is unclear what the right way is to do it. before there was the llvm-dev mailing list, which was okay-ish with answers, 50% of my issues I had were solved, but now that it is moved to the new forum this has dropped to 0%. I think the main devs just dont use it.

**Reply by TheCoryBarker** [reply]
Sorry for my slowness, I've been offline for a while to do sickness.

Thanks for the information!

If you remember I talked to you before when I put out a PR for some documentation [here](https://github.com/AFLplusplus/AFLplusplus/pull/1501), and now I want to start work on supporting CMPLOG for Android directly in Android build system. It doesn't seem like it would be too much work, but the maintenance might be concerning. I'm trying to get an idea of the level of maintenance effort I should be expecting for supporting CMPLOG in Android build system over time. The clang used in Android is ahead of the public LLVM releases, so the worry is that changes in LLVM API would affect Android before anyone else.

So between stuff getting deprecated, and unannounced "features" that break compilation, my question then is, as someone who's familiar with CMPLOG passes over time, how much effort do you guess it takes to keep these maintained on an average year?

I know this is kind of a ridiculous question and very hard to answer, but I don't have any clue here so any context you can provide is helpful.

**Reply by vanhauser-thc** [reply]
hard to answer. can be quiet for a year without anything to do, could be every few months.
but for us it would be great because it would give us early warning if something breaks.
and I can help you to get it working again.

**Reply by TheCoryBarker** [reply]
I see, that's very helpful. Thank you!
